# 01. Import libraries and data

In [2]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [3]:
# Define path to data

path = r'/Users/louise/Desktop/CF Coursework/Achievement 4 /Instacart Basket Analysis'

In [5]:
# Import products data

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
#Import wrangled orders data

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 02. Check data consistency

In [5]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
1                 1   2398795        1    prior             2          3   
2                 2    473747        1    prior             3          3   
3                 3   2254736        1    prior             4          4   
4                 4    431534        1    prior             5          4   
...             ...       ...      ...      ...           ...        ...   
3421078     3421078   2266710   206209    prior            10          5   
3421079     3421079   1854736   206209    prior            11          4   
3421080     3421080    626363   206209    prior            12          1   
3421081     3421081   2977660   206209    prior            13          1   
3421082     3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 8 columns]

In [7]:
# Get rid of unnamed column duplicating the order_id column

df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [8]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [8]:
# Check df_prods for null values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
# Create a subset of missing product names

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
# How many rows are in this dataframe? Is 16 a large portion? No

df_prods.shape

(49693, 5)

In [12]:
# Clean df_prods by removing the df_nan subset

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
# Have we removed the correct number of rows? Yes

df_prods_clean.shape

(49677, 5)

In [14]:
# Subset duplicated rows

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [15]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
# Drop duplicates from df_prods_clean

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
# Have we removed the correct number of rows? Yes

df_prods_clean_no_dups.shape

(49672, 5)

In [18]:
# Check major stats of df.prods using the describe function

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Comments: Product, aisle, and department ids all follow expected patterns. The only eybrow raiser is the max price: 99,999. This is probably a mistake (especially given that the third quartile is only 11.20).

In [19]:
# Check df_ords for mixed-type data; Found none

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [20]:
# Check df_ords for missing values

df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are a lot of missing values in the days_since_prior_order column! I will loose a lot of data if I simply remove 206,209 rows. I will use the desribe and shape functions on df_ords to see whether it is better to remove them of to impute the missing values.

In [21]:
df_ords.shape

(3421083, 7)

In [22]:
# Check what percentage of rows have missing values

206209 / 3421083

0.06027594185817766

Comments: Over 5%. I would like to avoid deleting all these rows if I can impute values fairly confidently. Let's look at the df_ords.desribe

In [23]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Comments: The recorded values in days_since_prior_order are spread about as would be expected, from 0 to 30 days. I think it is better to impute the missing values than to delete 206,209 rows. Let's compare the mean to the median.

In [24]:
df_ords['days_since_prior_order'].median()

7.0

Comments: The mean (11.11) is higher than the median (7), but falls within the thrid quartile and is not skewed by an extreme max value. I propose to go with the mean as the more accurate measure of the central tendency of the column. 

In [42]:
df_ords['days_since_prior_order'].fillna(11.11484, inplace = True)

In [43]:
# Check new major stats, missing values, and stats to see the change was a success

df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.421083e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            8.924952e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            5.000000e+00  
50%         1.300000e+01            8.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [44]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
observation_status        0
order_number              0
order_day_of_week         0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [45]:
df_ords.shape

(3421083, 7)

In [47]:
# Make a subset of any duplicate rows in df_ords

df_ords_dups = df_ords[df_ords.duplicated()]

In [49]:
df_ords_dups.shape

(0, 7)

Comments: Found none. Empty subset. No action is required. 

In [12]:
# Export df_ords as orders checked

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [51]:
# Export df_prods_clean_no_dups as products checked

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))